docs/README.md

```diff
@@ -0,0 +1,358 @@
+state from the stack adds new interval with corresponding state name into the event collection.
+
+It is worth noting there really are no semantics attached to states. The information that
+something is, for example, an OpenMP iteration period is solely stored in an instance
+of am_dsk_openmp_iteration_period. This data structure also contains a
+field for the interval. There is no mechanism to link a state event with
+anything else.
+
+More verbose types can be used to capture more information,
+however they have to be added to the type system. Extending type system
+will be covered later in this document.
+
+## Tracing
+
+### Initialization and destruction
+
+Main container for all information collected is `am_buffered_trace` object. It can
+directly store trace-wide data, e.g. types ids, global events, hierarchies, etc.
+Any trace with can be associated with any number of `am_buffered_event_collection`
+objects that allow more fine-grained storage of information, for example data
+associated with execution details of one thread.
+
+Before tracing starts the trace has to be initialized and written with ids of
+all used types. A this point new hierarchy can be also initialized (fully or
+partially) depends on the information available. At the end entry for each
+event collection has to be written to the trace using `am_dsk_event_mapping`
+object and `am_dsk_event_mapping_write_to_buffer_defid()` function and then trace
+has to be dumped to the file.
+
+To use event collection new event collection has to be initialized
+(`am_buffered_event_collection_init()`), added to the trace
+(`am_buffered_trace_add_collection()`) and written to disk
+(`am_dsk_event_collection_write_to_buffer_defid()`). Usually stack is also
+needed to track nested events and it is usually created at the same time
+as event collection. As there is no generic stack implementation in Aftermath
+it is up to the programmer to take care of it. Trace and event collections have
+no notion of stacks.
+
+Please see instrumented pthreads library or AfterOMPT for examples of creating
+and destroying traces.
+
+### Adding events to trace
+
+Assuming an appropriate type was added to the type system tracing events is
+relatively straightforward. For any event a new instance of `am_dsk_type_name` has
+to be created and then it can be written using `am_dsk_type_name_write_to_buffer_defid()`.
+The function takes as an argument data field of event collection and created event.
+
+## Type System
+
+### Adding new on-disk type
+
+To being able to trace events new type has to be added to type system for any new
+type. The types definitions can be found in `aftermath/libaftermath-core/src/defs/aftermath/types`
+and on-disk types are defined in `on_disk.py`. Adding new types consists of adding
+a new object of type `EventFrame` for event collections events and `Frame` for trace
+wide events. Then the event has to be added to `TypeList` at the end of the file.
```

Comment:
I added pointer to: libaftermath-core/doc/TYPE_SYSTEM since it's a good explanation and it's no point in duplicating work. This document is more of practical guide with the core part being step by step instruction for adding a new type (nb. I've been using it over and over again when adding types).